### PR TITLE
added test cases for recoverToken

### PR DESCRIPTION
**Pull request by xenide**
So I think I managed to write the test cases according to the specs given. Thanks for providing such a clear and detailed synopsis so I could get started quickly, I really appreciate the effort 🙏🏽 

Some additional comments:
1. To create the 3rd token (token to be recovered) dependency, I decided to just ride on the PairFixture for simplicity's sake. I considered the alternative of deploying the 3rd token contract within the test function itself, but I would need to import a bunch of other functions (like deployContract) into the test file, which seemed messy and anti-pattern. Another alternative is to create another fixture distinct from the PairFixture to support this. Any comments on this? 

2. I tried to follow the existing style, but I feel I'm still not getting it 100%. So please let me know what to change to adhere to the style better 

Thanks

### CODE REVIEW

**Comment by leodsoft** [comment]
Yeah - I think what you did for the additional token is fine in the fixture is the best approach.

On style etc, yes for this one we're just operating off the original uniswap precedent ... so its fine and your addition looks consistent which is the main goal. For future projects we will most likely make some big improvements (for the test and actual contract coding style).

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [inline uniswap-v2-core/test/UniswapV2Pair.spec.ts]: This is fine, would break if order of `require()` calls changed but the test case is sufficient for what we need

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [inline uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Using `AddressZero` is a little ambiguous here (because we want to revert due to `recover === 0`, not invalid token. Do you want to replace with `0x3704E657053C02411aA2Fd0599e75C3d817F81BC`? Ideally store `0x3704E657053C02411aA2Fd0599e75C3d817F81BC` in a `const` within the test file

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [inline uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Would this revert with `UniswapV2: TRANSFER_FAILED`?

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [inline uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Difference in `to.be.eq` and `to.eq`?

**Review by OliverNChalk — COMMENTED** [review]
Good coverage, have some minor comments

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (leodsoft) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: I believe they are equivalent and exist in an attempt at approximating more natural language in the code ... so its not a big deal but we should just standardise style wise - I think most of the code uses the shorter form, so we can just drop the 'be'.

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (leodsoft) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: yeah good pickup for completeness - although it will fail earlier for a separate reason, it would be better to not have an invalid input for the call we are using to test for failure of that separate reason.

There is though already a valid defined variable to use in that light: should probably just be recoverToken( token2.address ) as a valid token address?

Although redundant knowing the actual contract implementation, it may also be more complete to transfer some balance in for these types of tests are a setup step. (ie so in the end, this is testing an otherwise completely valid recover call, except for the zero recoverer address).

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (leodsoft) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Bringing @OliverNChalk' s comment above here, style wise, although it is commented making it clear, it's probably better to make this a constant with a clear name (INVALID_ADDRESS etc). 

And yes please add to specific revert message expected using revertedWith().  I think it will be "Vexchange: RECOVERER_ZERO_ADDRESS"

**Comment by leodsoft** [comment]
Note: I can resolve the conflicts before merging, after this branch is ready with any fixes for above.

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (xenide) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: you mean if we had swapped these two lines in `UniswapV2Pair.sol`, the test case wouldn't work? 

` require(token != token0, "Vexchange: INVALID_TOKEN_TO_RECOVER"); `
`require(token != token1, "Vexchange: INVALID_TOKEN_TO_RECOVER");` 

If I'm not mistaken, I think it still would

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (xenide) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: yes they're equivalent. I'll change it to the shorter form.

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (xenide) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: In my test setup the error I got for the invalid token address case (bottom one), was 
`RuntimeError: VM Exception while processing transaction: revert` 
So it's unlikely that it even reached line 108 of `UniswapV2Pair.sol` . But it's unclear at which line it failed. It could be line 262 in or it could be in line 107. Is there any way for me to test this? I would need to import the IERC20 interface into the test file and call the balanceOf function? 

also I noticed for the invalid token address test case that I forgot to set the recoverer's address in the first place. Cuz if the recoverer's address is not even set, the 
`require(recoverer != address(0), "Vexchange: RECOVERER_ZERO_ADDRESS");` 
would already fail before we even reach `_safeTransfer`. So I'll add in the code to set the recovery address.

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (xenide) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: yup it will be more consistent to store that invalid address in a const, so I'll do that. 

Regarding line 463, I agree that it'd be more appropriate to use a valid token address instead of AddressZero, so I'll change that. I'll also add the case of the account already having some tokens, while having recovery address zero.

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]:
```
require(recoverer != address(0), "Vexchange: RECOVERER_ZERO_ADDRESS");
```

Would trigger on this test case if it was the first require?

Comment (xenide) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: you're right, it would. So do you think I should craft the test case such that the pair has a recoverer address, so that we are only testing for the validity of the token address?

Comment (OliverNChalk) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Yes, while a bit pedantic, it is in-line with the goal of testing a single thing at a time

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Awesome, using token2 and having token2 in the contract will prove that we are just testing the recover zero :)

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Yeah, I'm not sure what the first error will be if we call a non-existant contract. I.e. if it's a valid address, then EVM will assume its EOA and thus my throw because we tried to call an EOA like a contract?

I'm okay not to invest a lot of time solving this issue because we would expect grabage in garbage out here

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (OliverNChalk) [inline uniswap-v2-core/test/UniswapV2Pair.spec.ts]: Prefer `const` when not re-assigning. This would usually be caught by a linter but im unsure if Uniswap have that in the repo

**Review by OliverNChalk — APPROVED** [review]
Looks good, I'm happy for this to be merged. You can do it yourself or we can leave it to Ian for first thing Monday because there are some conflicts to be resolved

File: uniswap-v2-core/test/UniswapV2Pair.spec.ts
Comment (xenide) [reply uniswap-v2-core/test/UniswapV2Pair.spec.ts]: okay I changed all appropriate instances to use `const` instead of `let`

**Review by leodsoft — APPROVED** [review]
Looks good.

For recoverToken:invalidToken ==> yes its properly focused on the recoverToken call now for the exception which is good, and it looks like it is difficult to try and catch errors from casting an invalid address... this type of thing could be a security risk in some cases, but should be safe for us here I believe.

**Pull request merged**
